### ISSUE TITLE
🐞 types.Map not supported at modules

### ISSUE THREAD

**Issue by aweris** [dagger/dagger#6138]
### What is the issue?

Currently, zenith do not support `map` types, resulting in compilation errors.

### Log output

```shell
dagger call workflow string
✘ load call ERROR [3.73s]
├ [3.73s] loading module
┃ Error: failed to get loaded module ID: input:1: host.directory.asModule.s
┃ ve failed to load module types: failed to call module "gale" to get funct
┃ ns: failed to get function output directory: process "/usr/local/bin/code
┃ n --module . --propagate-logs=true --introspection-json-path /schema.json
┃ did not complete successfully: exit code: 1
✘ serve ERROR [0.49s]
✘ exec /usr/local/bin/codegen --module . --propagate-logs=true --introspection-json-path /schema.json ERROR [0.47s]
┃ Error: generate code: template: module:59:3: executing "module" at <Modul
┃ ainSrc>: error calling ModuleMainSrc: failed to convert field type: unsup
┃ rted type *types.Map
✘ generating go module: gale ERROR [0.15s]
├ [0.03s] go mod tidy
┃ writing dagger.gen.go
┃ writing go.mod
┃ writing go.sum
┃ needs another pass...
• Cloud URL: https://dagger.cloud/runs/fbd1131a-6339-4d26-b2d1-98e5c6d533fd
• Engine: 4c1ca39d411b (version devel ())
⧗ 3.89s ✔ 170 ∅ 8 ✘ 4
```

### Steps to reproduce

Use `map[string]string` in the zenith module with golang SDK like following example 

```go
// Step holds the configuration of a step.
type Step struct {
	StepID string
	Name   string
	Env    map[string]string
	With   map[string]string
	Uses   string
}
```

and execute the module

### SDK version

Go SDK v0.9.3

### OS version

macOS 14.0

**Comment by helderco** [comment]
I don't think we can support this because GraphQL doesn't have maps. The only way is to have it "map" to a sequence of key/value pairs underneath. We should at least give a user friendly error here, explaining this.

**Comment by helderco** [comment]
In Go there's the additional issue of maps not preserving order, so that can affect caching.

**Comment by jedevc** [comment]
> I don't think we can support this because GraphQL doesn't have maps. We should at least give a user friendly error here, explaining this though.
> 
> The only way is to have it "map" to a sequence of key/value pairs underneath, like EnvVariable.

I think having the translation is the right call long term - not supporting maps is kinda frustrating, and I've definitely reimplemented the proposed solution myself several times when I've wanted kv pairs.

> In Go there's the additional issue of maps not preserving order, so that can affect caching.

Potentially, we could normalize and sort input. I don't think this is a concern for go (in the engine or SDK) since marshalling order of JSON is lexographically sorted.

**Comment by helderco** [comment]
> Potentially, we could normalize and sort input. I don't think this is a concern for go (in the engine or SDK) since marshalling maps order of JSON is lexographically sorted.

Yes, but I'm thinking about what you do with it inside the function. However you do have to power to use it in a way that doesn’t affect caching.

> > The only way is to have it "map" to a sequence of key/value pairs underneath, like EnvVariable.
> 
> I think having the translation is the right call long term - not supporting maps is kinda frustrating, and I've definitely reimplemented the proposed solution myself several times when I've wanted kv pairs.

To avoid having an input type for the argument and another for the output, we can basically make a general replacement for `EnvVariable`:

https://github.com/dagger/dagger/blob/aec81d10a55db52937a5e67e5ad8870b29fe9a35/docs/docs-graphql/schema.graphqls#L1128-L1138

That’s only for string values though. We’d need multiple versions of this to support other types.

So a function defined like this:

```go
func (m *Test) Sorted(env map[string]string) map[string]string {
    // sort env
    return env
}
```

Would produce:
```graphql
type Test {
    sorted(env: [KeyValueID!]!): [KeyValue!]!
}
```

Assuming: 
```graphql
scalar KeyValueID

type KeyValue {
    id: KeyValueID!
    name: String!
    value: String!
}

type Query {
    keyValue: KeyValue!
}
```

And in client code, used like this:

```go
dag.Test().Sorted(
    []*KeyValue{
        dag.KeyValue("name1", "value1"),
        dag.KeyValue("name2", "value2"),
    },
)
```

We could also try to have codegen handle this transparently so you never have to use `KeyValue` directly, but sometimes order does matter and automatically sorting may not produce the intended result. I guess you’d not use a `map[string]string` in this case.

```go
r, err := dag.Test.Sorted(ctx, map[string]string{
    "name1": "value1",
    "name2": "value2",
})
// r = map[name1:value1 name2:value2]
```

Or more simply, using a `KeyValue` scalar that uses the `name=value` format, so it could be used in both the input and output and when used in a list, translate it to a map in codegen.


Reminder that we had a similar request for using a map with a `WithEnvVariables` in:
- https://github.com/dagger/dagger/issues/4835
